### PR TITLE
Fixup covar2d shrinker and get_cwf_coeffs for  noise_var=0 

### LINKED ISSUE #420: Cov2d noise_var=0 appears to have a bug (crash)
There appears to be a discrepancy between 0 and 1 for the variable noise_var in cov2d family of code.  This is similar to an issue I found in the legacy code, but I was expecting it to be handled in ASPIRE-Python because it is specifically mentioned in the docstrings...  However, trying to use noise_var=0 with this code (for using clean images in development) results in a crash.

It would appear most of this cov2d code uses a default `noise_var=1`, I'm guessing to superficially circumvent this crashing in the shrinker codes.

To repro easiest thing is to alter the cov2d tutorial:
```
(aspire_cls_avg) ➜  ASPIRE-Python git:(cls_avg_cov2d) ✗ diff tutorials/examples/cov2d_bugcheck.py tutorials/examples/cov2d_simulation.py 
41c41
< noise_var = 0 #1.3957e-4
---
> noise_var = 1.3957e-4
```

bt:
```
(aspire_cls_avg) ➜  ASPIRE-Python git:(cls_avg_cov2d) ✗ python tutorials/examples/cov2d_bugcheck.py 
2021-05-25 11:50:27,700 INFO This script illustrates 2D covariance Wiener filtering functionality in ASPIRE package.
2021-05-25 11:50:27,700 INFO Simulation running in <class 'numpy.float32'> precision.
2021-05-25 11:50:27,700 INFO Initialize simulation object and CTF filters.
2021-05-25 11:50:27,700 INFO Load 3D map and downsample 3D map to desired grids of 64 x 64 x 64.
2021-05-25 11:50:28,162 INFO Use downsampled map to creat simulation object.
2021-05-25 11:50:28,181 INFO Appending a NoiseAdder to generation pipeline
2021-05-25 11:50:28,181 INFO Expanding 2D image in a frequency-domain Fourier–Bessel basis using the fast method.
2021-05-25 11:50:28,647 INFO Apply CTF filters to clean images.
2021-05-25 11:50:29,057 INFO Trying NFFT backend finufft
2021-05-25 11:50:29,060 INFO NFFT backend finufft usable.
2021-05-25 11:50:29,060 INFO Trying NFFT backend cufinufft
2021-05-25 11:50:29,060 INFO NFFT backend cufinufft not usable:
	No module named 'pycuda'
2021-05-25 11:50:29,060 INFO Trying NFFT backend pynfft
2021-05-25 11:50:29,061 INFO NFFT backend pynfft not usable:
	No module named 'pynfft'
2021-05-25 11:50:29,061 INFO Selected NFFT backend = finufft.
tutorials/examples/cov2d_bugcheck.py:97: RuntimeWarning: divide by zero encountered in double_scalars
  sn_ratio = power_clean / noise_var
2021-05-25 11:50:32,228 INFO Signal to noise ratio is inf.
2021-05-25 11:50:37,565 INFO Applying forward transformations in pipeline
2021-05-25 11:50:37,565 INFO All forward transformations applied
2021-05-25 11:50:37,565 INFO Loaded 1024 images
2021-05-25 11:50:37,565 INFO Get coefficients of clean and noisy images in FFB basis.
2021-05-25 11:50:40,000 INFO Get 2D covariance matrices of clean and noisy images using FB coefficients.
/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py:32: RuntimeWarning: divide by zero encountered in true_divide
  covar = covar_in / noise_var
2021-05-25 11:50:43,297 CRITICAL Array must not contain infs or NaNs
Traceback:
  File "tutorials/examples/cov2d_bugcheck.py", line 154, in <module>
    covar_est_opt=covar_opt,
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 305, in get_covar
    covar_est_opt["shrinker"],
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 372, in shrink_covar_backward
    b_ell = shrink_covar(b_ell, noise_var, p / n, shrinker)
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 34, in shrink_covar
    lambs, eig_vec = eig(make_symmat(covar))
  File "/home/garrett/anaconda3/envs/aspire_cls_avg/lib/python3.6/site-packages/numpy/linalg/linalg.py", line 1305, in eig
    _assertFinite(a)
  File "/home/garrett/anaconda3/envs/aspire_cls_avg/lib/python3.6/site-packages/numpy/linalg/linalg.py", line 218, in _assertFinite
    raise LinAlgError("Array must not contain infs or NaNs")

Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/exceptions.py", line 87, in handle_exception
    raise exc_value
  File "tutorials/examples/cov2d_bugcheck.py", line 154, in <module>
    covar_est_opt=covar_opt,
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 305, in get_covar
    covar_est_opt["shrinker"],
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 372, in shrink_covar_backward
    b_ell = shrink_covar(b_ell, noise_var, p / n, shrinker)
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 34, in shrink_covar
    lambs, eig_vec = eig(make_symmat(covar))
  File "/home/garrett/anaconda3/envs/aspire_cls_avg/lib/python3.6/site-packages/numpy/linalg/linalg.py", line 1305, in eig
    _assertFinite(a)
  File "/home/garrett/anaconda3/envs/aspire_cls_avg/lib/python3.6/site-packages/numpy/linalg/linalg.py", line 218, in _assertFinite
    raise LinAlgError("Array must not contain infs or NaNs")
numpy.linalg.LinAlgError: Array must not contain infs or NaNs

Original exception was:
Traceback (most recent call last):
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/exceptions.py", line 87, in handle_exception
    raise exc_value
  File "tutorials/examples/cov2d_bugcheck.py", line 154, in <module>
    covar_est_opt=covar_opt,
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 305, in get_covar
    covar_est_opt["shrinker"],
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 372, in shrink_covar_backward
    b_ell = shrink_covar(b_ell, noise_var, p / n, shrinker)
  File "/home/garrett/work/ClassAveraging/ASPIRE-Python/src/aspire/covariance/covar2d.py", line 34, in shrink_covar
    lambs, eig_vec = eig(make_symmat(covar))
  File "/home/garrett/anaconda3/envs/aspire_cls_avg/lib/python3.6/site-packages/numpy/linalg/linalg.py", line 1305, in eig
    _assertFinite(a)
  File "/home/garrett/anaconda3/envs/aspire_cls_avg/lib/python3.6/site-packages/numpy/linalg/linalg.py", line 218, in _assertFinite
    raise LinAlgError("Array must not contain infs or NaNs")
numpy.linalg.LinAlgError: Array must not contain infs or NaNs
```

Fixing this bug will look like:
- [x]  Reverse engineering what the code and value intent probably was.
- [x]  Patching the code.
- [x]  Patching the docstrings.
- [x]  Testing the case above.
- [x]  Confirming running clean images through the cov2d tool yields a sane result wrt the known clean images and opening a new issue if that is a larger problem.

### PR DESCRIPTION
**Pull request by garrettwrong**
Cleaned up a little so we no longer crash with div0 or singular matrices.  However, estimated result for clean case `get_cwf_coeffs` doesn't seem right (to me).

Maybe its too degenerate a case and my expectations are wrong?  If that is the case we can document, log, then no-op or something.. Essentially I'm hoping to be able to setup some code and turn noise on/off without effectively writing a totally new experiment...  We might need that for say integrating with class averaging or other developmental codes using clean images to start with...

Closes #420

### CODE REVIEW

**Comment by garrettwrong** [comment]
I'll close this and reopen seperate from the covar=0.  I would still like these cleanup changes though (should not have "None" and None, or default var=1 imo.

**Comment by garrettwrong** [comment]
I think the batched issue was because may be because were on the edge of tolerance in singles.  Probably the larger magnitudes (dividing by small noise_var) helped before.  It's possible there is more to it though...

The tutorial result, while still passable, appears to have more error now. I was expecting it to be tail noise, but its noticeable.  

For noise_var=0 I think its still not right (even with `get_cwf_coeffs` returning the coeffs identically now).  I get closer with negation, and I see a curious negative sign in the tutorial plots. Maybe its superfluous, but do you recall why that would be there?

**Comment by janden** [comment]
> I think the batched issue was because may be because were on the edge of tolerance in singles. Probably the larger magnitudes (dividing by small noise_var) helped before. It's possible there is more to it though...


Ah ok. I assumed it had something to do with the new default `noise_var`. It was using `noise_var = 1` before, which shouldn't have resulted in large magnitudes.
 
> The tutorial result, while still passable, appears to have more error now. I was expecting it to be tail noise, but its noticeable.

That's strange. Are you talking about with non-zero `noise_var`? It could be that I messed up the formulas for the shrinkage (although the relevant tests pass, so that shouldn't happen). What happens if you try without shrinkage?

> For noise_var=0 I think its still not right (even with `get_cwf_coeffs` returning the coeffs identically now). I get closer with negation, and I see a curious negative sign in the tutorial plots. Maybe its superfluous, but do you recall why that would be there?

There's bound to be some residual error due to projection onto the basis (I see some rings outside of the central disk), but otherwise it should be similar, especially if `get_cwf_coeffs` is set to pass through for `noise_var = 0`, since then it's just a matter of `evaluate_t` followed by `evaluate`.

**Comment by garrettwrong** [comment]
shrinker=None with the stock `noise_var` in the cov2d tutorial is closer (probably okay). Closer than `frobenius_norm`.

Can't compare the shrinker=None with `noise_var=0` because that case was previously not working.

I think in the tutorial I am seeing other things, maybe its okay. Can talk about it tomorrow. I'll try to have them ready to run and maybe can close this out then.

File: tests/test_covar2d.py [hunk 1]
```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
...
+        """
+        Test we raise when user supplies incorrect CTF arguments
+        """
+
+        with raises(RuntimeError):
+            _ = self.cov2d.get_cwf_coeffs(
+                self.coeff, self.h_ctf_fb, None, noise_var=self.noise_var
+            )
+
+    @parameterized.expand(
```
Comment (janden) [inline tests/test_covar2d.py]: `@pytest.mark.parametrize` and save us the dependency?

File: tests/test_covar2d.py [hunk 2]
```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
...
+        # Compare with clean images
+        delta = np.mean(np.square((self.imgs_clean - img_est).asnumpy()))
+        self.assertTrue(delta < 0.02)
+
+    def testGetCWFCoeffsCTFargs(self):
+        """
+        Test we raise when user supplies incorrect CTF arguments
+        """
+
+        with raises(RuntimeError):
```
Comment (janden) [inline tests/test_covar2d.py]: Check that error message is correct (`assert "x" in str(e.value)`).

File: tests/test_covar2d.py [hunk 3]
```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
...
+                self.coeff, self.h_ctf_fb, None, noise_var=self.noise_var
+            )
+
+    @parameterized.expand(
+        [(None,), "frobenius_norm", "operator_norm", "soft_threshold", "None"]
+    )
+    def testShrinkers(self, shrinker):
+        """Test all the shrinkers we know about run, and we raise otherwise."""
+
+        if shrinker == "None":
```
Comment (janden) [inline tests/test_covar2d.py]: Maybe another string would be better here (this just gives me flashbacks).

File: tests/test_covar2d.py [hunk 4]
```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
...
+            allow_pickle=True,
+        )
+
+        covar_coeff = self.cov2d.get_covar(
+            self.coeff_clean, covar_est_opt={"shrinker": shrinker}
+        )
+
+        for im, mat in enumerate(results.tolist()):
+            self.assertTrue(
+                np.allclose(mat, covar_coeff[im], atol=utest_tolerance(self.dtype))
```
Comment (janden) [inline tests/test_covar2d.py:221]: So since `noise_var = 0`, all these shrinkers should be no-ops, is that what we're testing?

File: tests/test_covar2d.py [hunk 5]
```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
...
+
+        coeff_cwf = self.cov2d.get_cwf_coeffs(
+            self.coeff_clean, self.h_ctf_fb, self.h_idx, noise_var=0
+        )
+
+        # Reconstruct images from output of get_cwf_coeffs
+        img_est = self.basis.evaluate(coeff_cwf)
+        # Compare with clean images
+        delta = np.mean(np.square((self.imgs_clean - img_est).asnumpy()))
+        self.assertTrue(delta < 0.02)
```
Comment (janden) [inline tests/test_covar2d.py:181]: Why such a large error? Due to projection error from FB basis?

**Review by janden — CHANGES_REQUESTED** [review]

File: tests/test_covar2d.py (hunk 4 above)
Comment (garrettwrong) [reply tests/test_covar2d.py:221]: I'm testing they should not crash.

File: tests/test_covar2d.py (hunk 1 above)
Comment (garrettwrong) [reply tests/test_covar2d.py]: AFAIK that doesn't work with unittest TestCase class... I can rewrite the tests to be functional instead of class based, but then we'll have two different styles of writing unit tests...

File: tests/test_covar2d.py (hunk 3 above)
Comment (garrettwrong) [reply tests/test_covar2d.py]: This was intentional. I will add another string.

File: tests/test_covar2d.py (hunk 5 above)
Comment (garrettwrong) [reply tests/test_covar2d.py:181]: No, I do not think it is from FB.  I think it is something with application and reversion of CTF or deeper... but fixing deeper issues is taking too long to be in the scope of this PR.

 I _think_ can use this as it is now.  It's up to you whether you want to stop other development to fix on cov2d more.

File: tests/test_covar2d.py (hunk 1 above)
Comment (janden) [reply tests/test_covar2d.py]: Ah that's too bad. Good to know.

File: tests/test_covar2d.py (hunk 4 above)
Comment (janden) [reply tests/test_covar2d.py:221]: Got it.

File: tests/test_covar2d.py (hunk 5 above)
Comment (janden) [reply tests/test_covar2d.py:181]: Oh ok. If you can isolate a small example, might be worth making an issue.

File: tests/test_covar2d.py (hunk 2 above)
Comment (garrettwrong) [reply tests/test_covar2d.py]: Sure, I used the `match` arg of `pytest.raises`.

File: tests/test_covar2d.py (hunk 3 above)
Comment (garrettwrong) [reply tests/test_covar2d.py]: I broke these out into two lists (`shrinkers` and `bad_shrinker_inputs`) at the class level so their intent is clearer.

While I was in there I also check the message returned in the exception, and added a note about removing `parameterized` if/when ASPIRE-Python moves away from the TestCase style.

File: tests/test_covar2d.py (hunk 5 above)
Comment (garrettwrong) [reply tests/test_covar2d.py:181]: For sure, I'd like to fix this too, but I got distracted too long already. (I really thought this corner case was simpler...).

#424

Comment (janden) [reply tests/test_covar2d.py:181]: Thanks.

File: tests/test_covar2d.py (hunk 2 above)
Comment (janden) [reply tests/test_covar2d.py]: Nice. I didn't know about that one. Much neater.

**Review by janden — APPROVED** [review]
